### PR TITLE
🐴Uniffe proposal 📰

### PR DESCRIPTION
**Pull request by samccone**
:sparkles:  **generated release diff** :sparkles: 

https://gist.github.com/samccone/237b7403edf1004ad969

xref #1309

If we like this I can extract the plugin or not... your call :dancer: 

If we hate this we can :skull: it

### CODE REVIEW

**Comment by jasonLaster** [comment]
Nice @samccone. Two questions
1. is there anything unique about `uniife` that would keep it from being it's own npm module. 
2. how is `uniffe` different from https://github.com/samccone/unwrap
3. can you comment the expected input and desired output in `uniffe` and the callee code because it can be a little hard to follow how the iife is dropped.

**Comment by samccone** [comment]
1. nope!
2. uniffe unwraps an iffe where unwrap unwraps a UMD wrapper (different structure)
3. yep

given

```js
(function() {
'use strict';

console.log('goat');
}());
```

uniffe results in 

```js
console.log('goat');
```

:golf:

File: uniffe.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
+'use strict';
+
+var through = require('through2');
+var escodegen = require('escodegen');
+var acorn = require('acorn');
+
+function uniffe(contents) {
```
Comment (surma) [inline uniffe.js]: Oh the irony: Should this be put in an IIFE?

Comment (samccone) [inline uniffe.js]: ha it is a node module so I think we are safe.... :goat:

**Comment by surma** [comment]
@samccone One remark, LGTM otherwise. 

I’d vote for landing this for now and exporting this as an stand-alone module later.

File: uniffe.js (hunk 1 above)
Comment (surma) [inline uniffe.js]: Alright :) Wasn’t sure about how this behaves in the wild.

**Comment by samccone** [comment]
if we want to land this, i can update the commit message to be something less "WIP"

**Comment by surma** [comment]
Microsite and templates work fine. `material.js` confirmed to only have one outer IIFE. 

Replace "WIP" with an emoji and this is good to land :pig: Going to wait for another ACK, though.

**Comment by surma** [comment]
@sgomes @Garbee @nicolasgarnier Can anyone of you take a look?

**Comment by Garbee** [comment]
I'm slightly confused on what the exact point is here.

It *appears* to me that it is taking all the individual component functions and making one big self-executing function for setting them up in-browser. Could be incorrect though, so please clarify if it is doing something else.

Just trying to make sure I understand the changes and their goal before passing judgement. :stuck_out_tongue:

**Comment by surma** [comment]
Completely fair, @Garbee.

We got complaints about global variables and lots of `'use strict';`. So #1307 was opened and I opened #1309 to wrap every component in a IIFE.

During #1309 we talked about that it is equally undesirable to have lots of IIFEs and would prefer one global IIFE (read discussion for details). So @samccone whipped up this which programmatically removes the component-specific IIFEs so we have as little code duplication as possible in the final output.

**Comment by samccone** [comment]
> we have as little code duplication as possible in the final output.

And we also have fewer unneeded closures :goat: :rocket:

**Comment by surma** [comment]
> And we also have fewer unneeded closures

Are you suggesting we care about memory? Oh right,… we do.

**Comment by Garbee** [comment]
Just did a quick manual check-over and it LGTM.

**Comment by surma** [comment]
Thanks, @Garbee :) Merging once @samccone has done the cleanup

**Comment by samccone** [comment]
cleaned up :tophat:

**Comment by surma** [comment]
No emoji in commit message.

p8ilx0e


Merging.

**Comment by samccone** [comment]
NOOOOOOO :cat:
